client/network/src/transport.rs

```diff
@@ -116,8 +119,19 @@ pub fn build_transport(
 		.upgrade(upgrade::Version::V1Lazy)
 		.authenticate(authentication_config)
 		.multiplex(multiplexing_config)
-		.timeout(Duration::from_secs(20))
-		.boxed();
+		.timeout(Duration::from_secs(20));
 
-	(transport, bandwidth)
+	if !memory_only {
+		let webrtc_transport = WebRTCTransport::new(keypair, webrtc_cert);
+
+		let transport = OrTransport::new(webrtc_transport, transport)
+			.map(|either_output, _| match either_output {
+				EitherOutput::First((peer_id, muxer)) => (peer_id, StreamMuxerBox::new(muxer)),
+				EitherOutput::Second((peer_id, muxer)) => (peer_id, StreamMuxerBox::new(muxer)),
+			})
+			.boxed();
```

Comment:
> A little bit offtopic, but what is the difference between OrTransport and EitherTransport in libp2p? Wasn't able to understand it from the docs.

`OrTransport` will first try 1st transport. If that fails with error (say `Multiaddr` is not supported), `OrTransport` will try the 2nd transport.

`EitherTransport` acts more like a https://docs.rs/either/latest/either/enum.Either.html, where you have two cases and `EitherTransport` will use one that maches (either Left or Right).

Reply:
So this update makes WebRTC transport preferable over plan TCP? I think it should be the opposite.

Reply:
> Wrapping muxer into StreamMuxerBox is not done above when combining TCP transport and WebSocket transport. Why is it needed here when combining with WebRTCTransport?

General info: TCP and WebSocket transports do not do multiplexing. In that way they can be easily combined. Yamux multiplexing is layered on top, transforming them into a stream muxer. In contrast, `WebRTCTransport` and `QUICTransport` both have streams and do "multiplexing in-house". So they are already muxers.

Answering your question, I had to wrap both into `StreamMuxerBox` so `Either` branches have the same type. Otherwise, you'd get:

```
    = note: expected tuple `(PeerId, libp2p::libp2p_webrtc::tokio::Connection)`
               found tuple `(PeerId, EitherOutput<Yamux<libp2p::libp2p_yamux::Incoming<multistream_select::negotiated::Negotiated<NoiseOutput<multistream_select::negotiated::Negotiated<BandwidthConnecLogging<EitherOutput<EitherOutput<EitherOutput<rw_stream_sink::RwStreamSink<BytesConnection<libp2p::libp2p_tcp::tokio::TcpStream>>, libp2p::libp2p_tcp::tokio::TcpStream>, EitherOutput<rw_stream_sink::RwStreamSink<BytesConnection<libp2p::libp2p_tcp::tokio::TcpStream>>, libp2p::libp2p_tcp::tokio::TcpStream>>, rw_stream_sink::RwStreamSink<Chan>>>>>>>>, libp2p::libp2p_mplex::Multiplex<multistream_select::negotiated::Negotiated<NoiseOutput<multistream_select::negotiated::Negotiated<BandwidthConnecLogging<EitherOutput<EitherOutput<EitherOutput<rw_stream_sink::RwStreamSink<BytesConnection<libp2p::libp2p_tcp::tokio::TcpStream>>, libp2p::libp2p_tcp::tokio::TcpStream>, EitherOutput<rw_stream_sink::RwStreamSink<BytesConnection<libp2p::libp2p_tcp::tokio::TcpStream>>, libp2p::libp2p_tcp::tokio::TcpStream>>, rw_stream_sink::RwStreamSink<Chan>>>>>>>>)`
```

Reply:
@nazar-pc good catch 👍 addressed in https://github.com/paritytech/substrate/pull/12529/commits/0db37d09f470f44805b19e981b5833c0e090898d